.github/workflows/dotnet-ec2-windows-test.yml

```diff
@@ -0,0 +1,334 @@
+            if [ $deployment_failed -eq 1 ]; then
+              echo "Destroying terraform"
+              terraform destroy -auto-approve \
+                -var="test_id=${{ env.TESTING_ID }}" 
+          
+              retry_counter=$(($retry_counter+1))
+            else
+              # If deployment succeeded, then exit the loop
+              break
+            fi
+          
+            if [ $retry_counter -eq $max_retry ]; then
+              echo "Max retry reached, failed to deploy terraform and connect to the endpoint. Exiting code"
+              exit 1
+            fi
+          done
+
+      - name: Get the ec2 instance ami id
+        working-directory: terraform/dotnet/ec2/windows
+        run: |
+          echo "EC2_INSTANCE_AMI=$(terraform output ec2_instance_ami)" >> $GITHUB_ENV
+
+      - name: Get SSM outputs
+        working-directory: terraform/dotnet/ec2/windows
+        run: |
+          echo "FRONTEND_SSM_ASSOCIATION_ID=$(terraform output frontend_script_association_id)" >> $GITHUB_ENV
+          echo "REMOTE_SSM_ASSOCIATION__ID=$(terraform output remote_script_association_id)" >> $GITHUB_ENV
+
+      - name: Get the sample app endpoint
+        working-directory: terraform/dotnet/ec2/windows
+        run: |
+          echo "MAIN_SERVICE_ENDPOINT=localhost:8080" >> $GITHUB_ENV
+          echo "REMOTE_SERVICE_IP=$(terraform output sample_app_remote_service_private_ip)" >> $GITHUB_ENV
+          echo "MAIN_SERVICE_INSTANCE_ID=$(terraform output main_service_instance_id)" >> $GITHUB_ENV
+          echo "REMOTE_SERVICE_INSTANCE_ID=$(terraform output remote_service_instance_id)" >> $GITHUB_ENV
+
+      - name: Monitor EC2 instances readiness in Systems Manager
+        run: |
+          max_attempts=60
+          attempt=0
+          interval=10
+
+          check_instance_ready() {
+            local instance_id=$1
+            while [[ $attempt -lt $max_attempts ]]; do
+              state=$(aws ssm describe-instance-information --filters Key=InstanceIds,Values=$instance_id --query "InstanceInformationList[0].PingStatus" --output text)
+              echo "Attempt $((attempt + 1))/$max_attempts: Current SSM state of instance $instance_id: $state"
+
+              if [[ "$state" == "Online" ]]; then
+                echo "Instance $instance_id is 'Online' in Systems Manager and ready."
+                break
+              else
+                echo "Waiting for instance $instance_id to be 'Online' in Systems Manager..."
+                sleep $interval
+                attempt=$((attempt + 1))
+              fi
+            done
+
+            if [[ $attempt -ge $max_attempts ]]; then
+              echo "Timeout reached: Instance $instance_id did not become 'Online' in Systems Manager within expected time."
+              exit 1
+            fi
+          }
+
+          check_instance_ready ${{ env.MAIN_SERVICE_INSTANCE_ID }}
+          check_instance_ready ${{ env.REMOTE_SERVICE_INSTANCE_ID }}
```

Comment:
`attempt` seems to be shared between the two `check_instance_ready()` calls, is this intended?

Reply:
yes, they are running in same time, so each one exceed max limit of time will treat as over time